Add group membership example to CCM api (#687)
chocolatey/docs: input/en-us/central-management/usage/api/examples.md

Register-ScheduledTask @taskParams
```

> ## Sync Central Management Groups with external data sources

Reviewer: ```suggestion
## Sync Chocolatey Central Management Groups With External Data Sources
```

> It is possible to keep your Central Management group(s) and their memberships up to date  with external data sources by leveraging the

Reviewer: ```suggestion
It is possible to keep your Chocolatey Central Management group(s), and their memberships, up to date with external data sources by leveraging the
```

Chocolatey Central Management API.

> The following functions will help you to use an external datasource such as SCCM, Active Directory, or a file to manage your Central Management groups!

Reviewer: ```suggestion
The following functions will help you to use an external datasource such as SCCM, Active Directory, or a file to manage your Chocolatey Central Management groups!
```

> ### Authenticating to the API

Reviewer: ```suggestion
### Authentication
```

> The first thing you'll need to do is authenticate to your Central Management instance with credentials that have appropriate permissions to manage group memberships. We'll use the following `Connect-CCMServer` function to provide that capability:

Reviewer: ```suggestion
The first thing you'll need to do is authenticate to your Chocolatey Central Management instance with credentials that have appropriate permissions to manage group memberships. We'll use the following `Connect-CCMServer` function to provide that capability:
```

```powershell
function Connect-CCMServer {
    [CmdletBinding()]
    Param(
        [Parameter(Mandatory)]
        [String]

>         $CcmServerHostname,

Reviewer: Should this not just be `Name` to be consistent with standard PowerShell functions / cmdlets? The function name indicates it's a server we're connecting to.

        [Parameter(Mandatory)]
        [System.Management.Automation.PSCredential]
        $Credential,

        [Parameter()]
        [Switch]
        $UseSSL
    )

    process {
        $script:$CcmHost = $CcmServerHostname
        $script:protocol = if($UseSSL) {
            'https'
        }
        else {
            'http'
        }

        $body = @{
            usernameOrEmailAddress = $Credential.UserName
            password               = $Credential.GetNetworkCredential().Password
        }
        $Result = Invoke-WebRequest -Uri "$($protocol)://$CcmServerHostname/Account/Login" -Method POST -ContentType 'application/x-www-form-urlencoded' -Body $body -SessionVariable Session -ErrorAction Stop

>         $script:Session = $Session

Reviewer: Should this not be `$global:Session = $Session`? I'm unsure why we set this to a script level variable?

Author: Same answer as before. Allows the CCM functions here to use it, but not be available elsewhere

    }
}
```

> This function accepts a `PSCredential` object via `-Credential` to store your credentials for Central Management. Additionally the `-CcmServerHostname` parameter accepts the DNS name you use to access Central Management's web interface. Use the `-UseSSL` switch parameter if you've configured your Central Management web interface to be accessed via an HTTPS connection (You _really_ should!).

Reviewer: ```suggestion
This function accepts a `PSCredential` object via `-Credential` to store your credentials for Chocolatey Central Management. Additionally the `-CcmServerHostname` parameter accepts the DNS name you use to access Central Management's web interface. Use the `-UseSSL` switch parameter if you've configured your Central Management web interface to be accessed via an HTTPS connection (You _really_ should!).
```

Reviewer: Should this not go into comments in the code itself? That way it is transported whenever the code is copied and will work for comment based function help?

#### Example

```powershell
Connect-CCMServer -CcmServerHostname centralmanagement.fabrikam.com -Credential (Get-Credential) -UseSSL
```

> ### Retrieving all computers currently in Central Management

Reviewer: ```suggestion
### Retrieving All Computers Currently In Chocolatey Central Management
```

> In order to accurately manage group membership we will need to retrieve all of the machines that are currently enrolled into Central Management. The following function will accomplish this.

Reviewer: ```suggestion
In order to accurately manage group membership we will need to retrieve all of the machines that are currently enrolled in Chocolatey Central Management. The following function will accomplish this:
```

```powershell
function Get-AllComputers {
    [CmdletBinding()]
    Param(
        [Parameter()]
        [Switch]
        $Detailed
    )

    begin {

>         if (-not $Session) { throw "Not connected to CCM instance. Run Connect-CCMServer first!" }

Reviewer: Should that not be `$global:Session` to indicate that this is coming from somewhere else? Or should it be explicitly passed in?

Reviewer 2: That only makes sense if this is distributed as a module. Which it's not, here, it's just a loose collection of functions, no? If folks are just copy-pasting this into a console, script: and global: are identical scopes.

    }

    process {
        $params = @{

>             Uri        = "$($protocol)://$CcmHost/api/services/app/Computers/GetAll"

Reviewer: This looks like it's assuming those variables exist, but they are not passed to the function. Chaining the functions together like this doesn't feel right. If we need anything inside the function, then this should be part of an overall script or module and call out specific `$script:` level variables so that you are not polluting the user session. Or it should be separate functions (which is what you have) that provide everything the function needs via parameters.

Author: They are set by Connect-CCMServer and will come along for the ride. Since we check for `$session`, also made by that function they'll be available. This keeps the function easier to use.

            Method     = 'GET'
            WebSession = $Session
        }
        $ComputerList = Invoke-RestMethod @params

> 

Reviewer: Do we not need to add any error checking here for invalid data?

        if (-not $Detailed) {
            $computerList.Result | Foreach-Object {
                [pscustomobject]@{
                    Id = $_.id
                    ComputerGUID = $_.computerGuid
                    Name         = $_.name
                    FriendlyName = $_.friendlyname
                    DisplayName  = $_.displayName
                    IpAddress    = $_.ipAddress
                }
            }
        }
        else {
            return $ComputerList.result
        }
    }
}
```

> This script works in conjunction with `Connect-CCMServer`. Failure to connect to your instance first will result in an error! This script includes a `-Detailed` switch parameter that can be used to include more detailed information about the machines reporting into your system should you want to leverage this function in other ways.

Reviewer: This kind of answers my question above, but there are two separate functions here. It's not a script.

> ### Validating a Group

Reviewer: ```suggestion
### Validating A Group
```

> We don't want to unnecessarily duplicate Groups inside of our Central Management instance. The following function sets up some guardrails

Reviewer: ```suggestion
We don't want to unnecessarily duplicate groups inside of our Chocolatey Central Management instance. The following function sets up some guardrails:
```

for us so that we prevent those duplicates.

```powershell
function Test-GroupExists {
    [CmdletBinding()]
    Param(
        [Parameter(Mandatory)]
        [String]
        $Group
    )

    begin {
        if (-not $Session) { throw "Not connected to CCM instance. Run Connect-CCMServer first!" }
    }

    process {
        $params = @{
            Uri = "$($protocol)://$CcmHost/api/services/app/Groups/GetAll"
            Method = 'GET'
            ContentType = 'application/json'
            WebSession = $Session
        }
        $result = Invoke-RestMethod @params

        $name = $result.result.name

        if ($Group -in $name) {
            return $true
        }
        else { return $false }

    }
}
```

> This function queries your Central Management instance for the group provided via the mandatory `-Group` parameter and returns `$true` or `$false` based on the group's existence in your environment.

Reviewer: ```suggestion
This function queries your Chocolatey Central Management instance for the group provided via the mandatory `-Group` parameter and returns `$true` or `$false` based on the group's existence.
```

> ### Adding a Group

Reviewer: ```suggestion
### Adding A Group
```

> Once we have connected to our environment, retrieved all our reporting computers, and validated our group we may need to actually _create_ the Group we want. The following function does just that!

Reviewer: ```suggestion
Once we have connected to our Chocolatey Central Management server, retrieved all our reporting computers, and validated our groups, we may need to actually _create_ the group we want. The following function does just that!
```

```powershell
function Add-CCMGroup {
    [CmdletBinding()]
    Param(
        [Parameter(Mandatory)]
        [String]
        $Group,

        [Parameter()]
        [String]
        $Description
    )

    begin {
        if (-not $Session) { throw "Not connected to CCM instance. Run Connect-CCMServer first!" }
    }

    process {
        if (-not (Test-GroupExists -Group $Group)) {
            $Params = @{
                Uri         = "$($protocol)://$CcmServerHostname/api/services/app/Groups/CreateOrEdit"
                Method      = 'POST'
                WebSession  = $Session
                ContentType = 'application/json'
                Body        = @{
                    name        = $Group
                    description = $Description
                    groups      = @()
                    computers   = @()
                } | ConvertTo-Json
            }
            $null = Invoke-RestMethod @params

            $params = @{
                Uri        = "$($protocol)://$CcmServerHostname/api/services/app/Groups/GetAll"
                Method     = "GET"
                WebSession = $Session
            }

>             [pscustomObject]$Group = Invoke-RestMethod @params |
>             Select-Object -ExpandProperty result |
>             Where-Object Name -EQ $Group | Select-Object Name, Id
> 
> 
>             return $Group

Reviewer: ```suggestion
            [pscustomObject](Invoke-RestMethod @params |
            Select-Object -ExpandProperty result |
            Where-Object Name -EQ $Group | Select-Object Name, Id)
```

Reviewer: Same result, just a suggestion to condense the code.

        }
        else {
            Write-Warning "Group already exists, skipping creation!"
        }
    }
}
```

> This function accepts the mandatory `-Group` parameter for the Group that we would like to create, and an optional `-Desciption` parameter if we would like to add a description to our newly created Group.

Reviewer: ```suggestion
This function accepts the mandatory `-Group` parameter for the group that we would like to create, and an optional `-Desciption` parameter if we would like to add a description to our newly created group.
```

> ### Managing the Group's membership

Reviewer: ```suggestion
### Managing Group Memberships
```

> Whew, getting closer! We've now got our functions for:

Reviewer: ```suggestion
Phew, getting closer! We've now got our functions for:
```

> - Connecting to CCM

Reviewer: ```suggestion
- Connecting to Chocolatey Central Management
```

- Getting all of our reporting computers

> - Verifying a Group's existence

Reviewer: ```suggestion
- Verifying a group's existence
```

> - Creating a new Group

Reviewer: ```suggestion
- Creating a new group
```

> This function will actually manipulate the members of our Group.

Reviewer: ```suggestion
This function will actually manipulate the members of our group:
```

```powershell
function Add-CCMGroupMember {
    [CmdletBinding()]
    Param(
        [Parameter(Mandatory)]
        [String]
        $Group,

        [Parameter(Mandatory)]
        [String[]]
        $Member
    )

    begin {
        if (-not $Session) { throw "Not connected to CCM instance. Run Connect-CCMServer first!" }
    }

    process {
        $CentralManagementComputers = Get-AllComputers

        #Get the ID of the group via CCM API
        $params = @{
            Uri        = "$($protocol)://$CcmHost/api/services/app/Groups/GetAll"
            Method     = "GET"
            WebSession = $Session
        }

        $GroupData = Invoke-RestMethod @params |
        Select-Object -ExpandProperty result |
        Where-Object Name -EQ $Group | Select-Object Name, Id
        $Id = $GroupData.Id
        $ccmGroupName = $GroupData.Name

        #Build a collection for the computerID payload
        $collection = [System.Collections.Generic.List[psobject]]::new()

>         #Loop SCCM Computers, and build object for Group payload into CCM API

Reviewer: ```suggestion
        #Loop CCM Computers, and build object for group payload into CCM API
```

                    computerId = $($CentralManagementComputers | Where-Object { $_.Name -eq $name}  | Select-Object -ExpandProperty Id)
                }

                $collection.Add($c)
            } else {
                Write-Warning "Computer: $_ is NOT currently reporting into Chocolatey Central Management!!"
            }
        }

        #Modify group membership via CCM API
        $params = @{
            Uri         = "$($protocol)://$CcmHost/api/services/app/Groups/CreateOrEdit"
            Method      = "POST"
            WebSession  = $Session
            ContentType = 'application/json'
            Body        = @{
                id        = $Id
                name = $ccmGroupName
                computers = $collection
            } | ConvertTo-Json
        }

        Write-Verbose $params.Body

        try {
            $null = Invoke-RestMethod @params -ErrorAction Stop
        }
        catch {
            $_.Exception.Message
        }
    }
}
```

This function takes a mandatory `-Group` parameter for the group you wish to manipulate as well as a mandatory `-Member` parameter which accepts an array of computer names.

> ### Put it all together

Reviewer: ```suggestion
### Putting It All Together
```

> #### Example - SSCM Collections

Reviewer: ```suggestion
#### Example - SCCM Collections
```

> This example is rather simple and assumes you have copied the above functions into your PowerShell session so they are available.

Reviewer: ```suggestion
This example reads data from SCCM and creates a Chocolatey Central Management group. It's assumes you have copied the above functions into your PowerShell session so they are available.
```

```powershell
Connect-CCMServer -CcmServerHostname centralmanagment.fabrikam.com -Credential (Get-Credential) -UseSSL
$Group = 'Shop Floor Computers'
$Members = Get-CMCollection -Name $Group | Get-CMCollectionMember | select-Object -ExpandProperty Name
Add-CCMGroupMember -Group $Group -Member $Members
```

#### Example - Active Directory

> This example gets the computers from an Active Directory OU and creates a Central Management group

Reviewer: ```suggestion
This example gets the computers from an Active Directory OU and creates a Chocolatey Central Management group.
```

```powershell
Connect-CCMServer -CcmServerHostname centralmanagment.fabrikam.com -Credential (Get-Credential) -UseSSL
$Group = 'Shop Floor Computers'
$Members = Get-ADComputer -Searchbase "OU=Floor,OU=Manufacturing,OU=Tuscon,DC=fabrikam,DC=com" -Properties Name | Select-Object -Expand Name
Add-CCMGroupMember -Group $Group -Member $Members
```

#### Example - CSV File

> This example reads data from a CSV file and creates a Central Management group

Reviewer: ```suggestion
This example reads data from a CSV file and creates a Chocolatey Central Management group.
```

        .EXAMPLE

        Connect-CCMServer -Name centralmanagement.fabrikam.com -Credential (Get-Credential) -UseSSL
    #>
    [CmdletBinding()]
    Param(
        #The FQDN of your Chocolatey Central Management instance.
        [Parameter(Mandatory)]
        [String]
        $Name,

        #Credentials for connecting to your Chocolatey Central Management instance
        [Parameter(Mandatory)]
        [System.Management.Automation.PSCredential]
        $Credential,

        #Use if you connect to Chocolatey Central Management over HTTPS
        [Parameter()]
        [Switch]
        $UseSSL
    )

    process {
        $script:CcmHost = $Name
        $script:protocol = if($UseSSL) {
            'https'
        }
        else {
            'http'
        }

        $body = @{
            usernameOrEmailAddress = $Credential.UserName
            password               = $Credential.GetNetworkCredential().Password
        }
        $Result = Invoke-WebRequest -Uri "$($protocol)://$Name/Account/Login" -Method POST -ContentType 'application/x-www-form-urlencoded' -Body $body -SessionVariable Session -ErrorAction Stop

>         $script:Session = $Session

Author: Shoot, good catch. Fixed it, waiting for remaining review for more changes before it is pushed up.
